### ISSUE TITLE
Add true async stream implementations for use in aspnetcore3+

### ISSUE THREAD

**Issue by kylemc** [MiloszKrajewski/K4os.Compression.LZ4#34]
I recently attempted use to the `LZ4EncoderStream` with aspnetcore3.1 by writing a custom `ICompressionProvider`. It worked great until I started using it in services with `AllowSynchronousIO` disabled.

https://github.com/dotnet/aspnetcore/issues/7644

At that point, I got the following exception.

```
"Message":"Synchronous operations are disallowed. Call WriteAsync or set AllowSynchronousIO to true instead.",
"StackTrace":"   at Microsoft.AspNetCore.Server.IIS.Core.HttpResponseStream.Write(Byte[] buffer, Int32 offset, Int32 count)\r\n
   at Microsoft.AspNetCore.Server.IIS.Core.WrappingStream.Write(Byte[] buffer, Int32 offset, Int32 count)\r\n
   at K4os.Compression.LZ4.Streams.LZ4EncoderStream.FlushStash()\r\n
   at K4os.Compression.LZ4.Streams.LZ4EncoderStream.WriteFrame()\r\n
   at K4os.Compression.LZ4.Streams.LZ4EncoderStream.Write(Byte[] buffer, Int32 offset, Int32 count)\r\n  
   at System.IO.Stream.<>c.<BeginWriteInternal>b__51_0(Object <p0>)\r\n
   at System.Threading.Tasks.Task`1.InnerInvoke()\r\n
   at System.Threading.Tasks.Task.<>c.<.cctor>b__274_0(Object obj)\r\n
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)\r\n
   --- End of stack trace from previous location where exception was thrown ---\r\n
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)\r\n
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)\r\n
   --- End of stack trace from previous location where exception was thrown ---\r\n
   at Microsoft.AspNetCore.ResponseCompression.ResponseCompressionBody.WriteAsync(Byte[] buffer, Int32 offset, Int32 count, CancellationToken cancellationToken)
```

It would be great if you could write stream implementations that are compatible with this new asp.net configuration option.

**Comment by kylemc** [comment]
I wrote up a straightforward `AsyncLZ4EncoderStream` implementation that duplicated every "write" method with an async pair. It has to ship on netstandard2.1+ because that's the first runtime to include `Stream.DisposeAsync`.

I have a simple unit test that runs on the aspnetcore test harness to verify correctness. I'll have to check if I'm able to contribute any of it back.

**Comment by MiloszKrajewski** [comment]
Please do, although I would need some conditional compilation as lot of people (including me) are still on netstandard 2.0, netstandard 1.6, Mono, Unity, and .NET 4.5.
I also wanted to actually rewrite this part to extract as much reused code for Sync and Async.

**Comment by MiloszKrajewski** [comment]
Addressed in 1.2.5
